File: core/src/xtdb/hyper_log_log.clj
(ns xtdb.hyper-log-log
  (:require [xtdb.memory :as mem])
  (:import [org.agrona DirectBuffer MutableDirectBuffer]
           java.nio.ByteOrder))

;; http://dimacs.rutgers.edu/~graham/pubs/papers/cacm-sketch.pdf
;; http://algo.inria.fr/flajolet/Publications/FlFuGaMe07.pdf

(def ^{:tag 'int} default-buffer-size (* Integer/BYTES 1024))

(defn add ^org.agrona.MutableDirectBuffer [^MutableDirectBuffer hll v]
  (let [m (quot (.capacity hll) Integer/BYTES)
        b (Integer/numberOfTrailingZeros m)
        x (mix-collection-hash (clojure.lang.Util/hasheq v) 0)
        j (bit-and (bit-shift-right x (- Integer/SIZE b)) (dec m))
        w (bit-and x (dec (bit-shift-left 1 (- Integer/SIZE b))))]
    (doto hll
      (.putInt (* j Integer/BYTES)
               (max (.getInt hll (* j Integer/BYTES) ByteOrder/BIG_ENDIAN)
                    (- (inc (Integer/numberOfLeadingZeros w)) b))
               ByteOrder/BIG_ENDIAN))))

(defn estimate ^double [^DirectBuffer hll]
  (let [m (quot (.capacity hll) Integer/BYTES)
        z (/ 1.0 (double (loop [n 0
                                acc 0.0]
                           (if (< n (.capacity hll))
                             (recur (+ n Integer/BYTES)
                                    (+ acc (Math/pow 2.0 (- (.getInt hll n ByteOrder/BIG_ENDIAN)))))
                             acc))))
        am (/ 0.7213 (inc (/ 1.079 m)))
        e (* am (Math/pow m 2.0) z)]
    (cond
      (<= e (* (double (/ 5 2)) m))
      (let [v (long (loop [n 0
                           acc 0]
                      (if (< n (.capacity hll))
                        (recur (+ n Integer/BYTES)
                               (+ acc (if (zero? (.getInt hll n ByteOrder/BIG_ENDIAN))
                                        1
                                        0)))
                        acc)))]
        (if (zero? v)
          e
          (* m (Math/log (/ m v)))))
      (> e (* (double (/ 1 30)) (Integer/toUnsignedLong -1)))
      (* (Math/pow -2.0 32)
         (Math/log (- 1 (/ e (Integer/toUnsignedLong -1)))))
      :else
      e)))

(defn combine ^DirectBuffer [^DirectBuffer hll-a ^DirectBuffer hll-b]
  (assert (= (.capacity hll-a) (.capacity hll-b)))
  (loop [n 0
         result ^MutableDirectBuffer (mem/allocate-unpooled-buffer (.capacity hll-a))]
    (if (= n (.capacity result))
      result
      (recur (+ n Integer/BYTES)
             (doto result
               (.putInt n (max (.getInt hll-a n ByteOrder/BIG_ENDIAN)
                               (.getInt hll-b n ByteOrder/BIG_ENDIAN)) ByteOrder/BIG_ENDIAN))))))

(defn estimate-union ^double [^DirectBuffer hll-a ^DirectBuffer hll-b]
  (estimate (combine hll-a hll-b)))

(defn estimate-intersection ^double [^DirectBuffer hll-a ^DirectBuffer hll-b]
  (- (+ (estimate hll-a)
        (estimate hll-b))
     (estimate-union hll-a hll-b)))

(defn estimate-difference ^double [^DirectBuffer hll-a ^DirectBuffer hll-b]
  (- (estimate hll-a)
     (estimate-intersection hll-a hll-b)))

File: core/src/main/kotlin/xtdb/util/HyperLogLog.kt
@file:JvmName("HyperLogLog")
package xtdb.util

import io.netty.buffer.ByteBuf
import io.netty.buffer.Unpooled
import xtdb.arrow.VectorReader
import java.nio.ByteBuffer
import kotlin.math.ln
import kotlin.math.max
import kotlin.math.pow

typealias HLL = ByteBuffer

// http://dimacs.rutgers.edu/~graham/pubs/papers/cacm-sketch.pdf
// http://algo.inria.fr/flajolet/Publications/FlFuGaMe07.pdf

const val DEFAULT_BUFFER_SIZE: Int = Integer.BYTES * 1024

operator fun HLL.plus(hash: Int): HLL {
    val m = this.capacity() / Integer.BYTES
    val b = Integer.numberOfTrailingZeros(m)


    val j = (hash ushr (Integer.SIZE - b)) and (m - 1)
    val w = hash and ((1 shl (Integer.SIZE - b)) - 1)

    val position = j * Integer.BYTES
    val currentValue = this.getInt(position)
    val newValue = max(currentValue, (Integer.numberOfLeadingZeros(w) + 1) - b)

    this.putInt(position, newValue)

    return this
}

fun HLL.add(vec: VectorReader, idx: Int, hasher: Hasher = Hasher.Xx()) = this + vec.hashCode(idx, hasher)

private val TWO_TO_32 = 1L shl 32
private val LARGEST_UNSIGNED_INT : Int = (TWO_TO_32 - 1).toInt()

fun HLL.estimate(): Double {
    val m = this.capacity() / Integer.BYTES

    var acc = 0.0
    for (i in 0 until this.capacity() step Integer.BYTES) {
        acc += 2.0.pow(-this.getInt(i))
    }

    val z = 1.0 / acc
    val alpha = 0.7213 / (1.0 + 1.079 / m)
    val e = alpha * m.toDouble().pow(2.0) * z

    // Small range correction
    if (e <= 2.5 * m) {
        var v = 0L
        for (i in 0 until this.capacity() step Integer.BYTES) {
            if (this.getInt(i) == 0) {
                v++
            }
        }

        return if (v == 0L) {
            e
        } else {
            m * ln(m.toDouble() / v)
        }
    }

    if (e > LARGEST_UNSIGNED_INT / 30.0 ) {
        return TWO_TO_32 * ln(1.0 - e / LARGEST_UNSIGNED_INT)
    }

    return e
}

fun HLL.combine(hll: HLL): HLL {
    require(this.capacity() == hll.capacity()) { "HyperLogLog buffers must have the same capacity" }

    val result = ByteBuffer.allocate(this.capacity())
    for (i in 0 until hll.capacity() step Integer.BYTES) {
        result.putInt(i, max(this.getInt(i), hll.getInt(i)))
    }
    return result
}

fun HLL.estimateUnion(hll: HLL): Double = combine(hll).estimate()

fun HLL.estimateIntersection(hll: HLL): Double = estimate() + hll.estimate() - estimateUnion(hll)

fun HLL.estimateDifference(hll: HLL): Double = estimate() - estimateIntersection(hll)

fun createHLL(bufferSize: Int = DEFAULT_BUFFER_SIZE): HLL = ByteBuffer.allocate(bufferSize)

fun toHLL(bytes: ByteArray): HLL = ByteBuffer.wrap(bytes)